rascal-lsp/src/main/rascal/lang/rascal/tests/rename/Functions.rsc

```diff
@@ -0,0 +1,212 @@
+    'int size(list[&T] _: [_, *l]) = 1 + size(l);
+", cursorAtOldNameOccurrence = -1, oldName = "size", newName = "sizeof");
+
+test bool patternOverloadFromDef() = {0, 1, 2, 3} == testRenameOccurrences("x = size([1, 2]);", decls = "
+    'int size(list[&T] _: []) = 0;
+    'int size(list[&T] _: [_, *l]) = 1 + size(l);
+", oldName = "size", newName = "sizeof");
+
+test bool typeOverloadFromUse() = {0, 1, 2, 3, 4, 5, 6} == testRenameOccurrences("x = size([1, 2]);", decls = "
+    'int size(list[&T] _: []) = 0;
+    'int size(list[&T] _: [_, *l]) = 1 + size(l);
+    '
+    'int size(set[&T] _: {}) = 0;
+    'int size(set[&T] _: {_, *s}) = 1 + size(s);
+", cursorAtOldNameOccurrence = -1, oldName = "size", newName = "sizeof");
+
+test bool typeOverloadFromDef() = {0, 1, 2, 3, 4, 5, 6} == testRenameOccurrences("x = size([1, 2]);", decls = "
+    'int size(list[&T] _: []) = 0;
+    'int size(list[&T] _: [_, *l]) = 1 + size(l);
+    '
+    'int size(set[&T] _: {}) = 0;
+    'int size(set[&T] _: {_, *s}) = 1 + size(s);
+", oldName = "size", newName = "sizeof");
+
+test bool arityOverloadFromUse() = {0, 1, 2, 3, 4, 5} == testRenameOccurrences("x = concat(\"foo\", \"bar\");", decls = "
+    'str concat(str s) = s;
+    'str concat(str s1, str s2) = s1 + concat(s2);
+    'str concat(str s1, str s2, str s3) = s1 + concat(s2, s3);
+", cursorAtOldNameOccurrence = -1, oldName = "concat", newName = "foo");
+
+test bool arityOverloadFromDef() = {0, 1, 2, 3, 4, 5} == testRenameOccurrences("x = concat(\"foo\", \"bar\");", decls = "
+    'str concat(str s) = s;
+    'str concat(str s1, str s2) = s1 + concat(s2);
+    'str concat(str s1, str s2, str s3) = s1 + concat(s2, s3);
+", oldName = "concat", newName = "foo");
+
+test bool overloadClash() = {0} == testRenameOccurrences("", decls = "
+    'int foo = 0;
+    'int foo(int x) = x when x \< 2;
+");
+
+test bool crossModuleOverload() = testRenameOccurrences({
+    byText("Str", "
+        'str concat(str s) = s;
+        'str concat(str s1, str s2) = s1 + concat(s2);
+    ", {0, 1, 2})
+    , byText("Main", "
+        'extend Str;
+        'str concat(str s1, str s2, str s3) = s1 + concat(s2, s3);
+    ", {0, 1})
+}, <"Main", "concat", 0>, newName = "conc");
+
+test bool simpleTypeParams() = {0, 1} == testRenameOccurrences("
+    '&T foo(&T l) = l;
+", oldName = "T", newName = "U");
+
+test bool typeParamsFromReturn() = {0, 1, 2} == testRenameOccurrences("
+    '&T foo(&T l) {
+    '   &T m = l;
+    '   return m;
+    '}
+", oldName = "T", newName = "U");
+
+test bool keywordTypeParamFromReturn() = {0, 1, 2} == testRenameOccurrences("
+    '&T foo(&T \<: int l, &T \<: int kw = 1) = l + kw;
+", oldName = "T", newName = "U");
+
+test bool typeParamsFromFormal() = {0, 1, 2} == testRenameOccurrences("
+    '&T foo(&T l) {
+    '   &T m = l;
+    '   return m;
+    '}
+", oldName = "T", newName = "U", cursorAtOldNameOccurrence = 1);
+
+test bool typeParamsClash() = {0, 1} == testRenameOccurrences("
```

Comment:
Isn't this renaming illegal? It changes the function signature, doesn't it?

Reply:
I probably missed an illegal rename expectation there. This is quite the can of worms...

Reply:
Fixed the test, added some more, and fixed the implementation.